Feat(IconDefault): retrieve all options from CSS

ghybs opened the pull request:
Should fix #4968 

This PR attempts to improve the Default Icon automatic images path resolution, especially for build engines (typically webpack) which can be configured to manage images linked through image CSS data types (e.g. `url(path/to/image.png)`) and re-write them, hence breaking the Leaflet path detection and re-building (the previously hard coded `marker-icon.png` file name no longer exists after the build engine renamed it or inlined it as base64 data).

To achieve this, this PR takes the radical approach to outsource **all** Default Icon option values to the CSS (image url's including file name, sizes and anchor positions, for icon, shadow and their Retina version) and retrieve all these values the first time `_getIconUrl` is called.

That way, everything related to the default icon images is located in one place (the CSS) and can be easily modified by the app developer (and the build engine…)

Due to this shift, `IconDefault` no longer has default options (except for a new option `classNamePrefix`, more on that below), and when retrieved from CSS, the option values are applied to the _instance_ options rather than on `IconDefault.prototype.options`.
That way, it makes it possible to use other instances of `IconDefault` with a different `classNamePrefix` option to target another set of CSS rules and define a new icon through that CSS. May not be very useful outside the default icon, but still an extra functionality.

There is however a change in how `imagePath` is applied: it must be set _before_ the first call to `_getIconUrl` occurs (typically before any Marker is added to a map), because it is no longer used after that. Whereas previously it can be changed dynamically and affect re-drawn Markers.

Another tweak is the `margin` (top and left) used to specify the icon (and shadow) anchor: to be closer to how currently we can position an icon through its `className` with negative margins, I took the approach to use negative margins here as well, although the `iconAnchor` (and `shadowAnchor`) typically have positive values. Then the margin values are reverted to get the positive options. (but then Leaflet re-reverts them to negative margins to make the actual positioning…)

Finally, I replaced the use of `background-image` by `cursor` CSS rule in order to workaround the high contrast / colours overriding mode in Firefox (see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-328277840).

Note that both the normal url and its Retina version are specified as multiple images in this `cursor` rule, as allowed by the spec.

Tests:
- Added 1 test for `imagePath`.
- Adjusted Popup and Marker tests to account for the new IconDefault behaviour.

Quite some code for this simple feature, so I am expecting some critics in review(s)… :-)

fluffywaffles commented:
@perliedman might you have time to review this soon? Ran into this issue again today; having the fix merged would be great!

cherniavskii commented:
@ghybs I remembered a bug I had once with a [cursor image in IE11](https://stackoverflow.com/questions/12454770/are-cursor-urls-relative-to-the-css-file).
Look like your solution may not work, if leaflet stylesheet will be placed not in site's root directory.

ghybs commented:
@cherniavskii thank you for the link!

Then it is impossible to satisfy both IE11 and Firefox high contrast option, unless we duplicate the CSS rule (having both `background` and `cursor`) so that we can fallback to `cursor` in case the former is empty...

Sigh even more bytes to add...

perliedman commented:
@ghybs thanks for your thoughtful work and comments regarding this feature! Also sorry about taking forever to review this.

I think I generally like this - I think this approach makes a lot of sense given the issues we've had. The major downside I can see is probably that it's not entirely intuitive to figure out which CSS property that corresponds to which icon option, but I don't think we can do much more than just document the fact.

My major concern now would be backwards compatibility: if you _do_ set icon options using the old way, will that still work? From just looking at the code, it looks like it wouldn't, and perhaps that's something we should address first.

ghybs commented:
Hi @perliedman,

Thank you for your feedback!

As for backwards compatibility with setting default icon options, I think it should work, even though there is an extra condition. I have added a test spec for that.

Given the fact that with this PR, options are applied _per instance_ of `L.Icon.Default` instead of being always retrieved from `L.Icon.Default.prototype`, these app specific options must be applied to `L.Icon.Default.prototype` **before** the first default icon tries to retrieve options from CSS; i.e. typically before any marker is added to a map.

However I think this is how apps should work already (i.e. setting default option _before_ adding markers to map), otherwise the result would be strange: any marker already added to map would use the previous options, and any _new_ marker, or those removed then re-added, would use the new options.

That being said, I think it should be possible to refactor this PR so that options are no longer applied per instance, but applied on `L.Icon.Default.prototype` like the current situation. But like the above example case shows, I think having per instance options is more logical.
Please let me know your opinion on this.

I still need to refactor to use `background-image` first for IE, then fallback on `cursor` for Firefox high contrast setting.

perliedman commented:
@ghybs thanks, that sounds very reasonable!

ghybs commented:
Refactored this PR to use `background-image` first (to avoid [IE bug](https://github.com/Leaflet/Leaflet/pull/5771#issuecomment-333618629) regarding `cursor` URL being relative to HTML page instead of relative CSS file) and fallback to `cursor` (for Firefox high contrast option).

Added a test spec for this fallback mechanism.

ghybs commented:
Unfortunately I think I still missed something, somehow implied by @perliedman's previous comment.

Current situation:
- _Common_ images folder retrieved from CSS. Recorded in `options.imagePath`, unless already provided.
- Icon URL option in JS class provides file name (but it could also include some path).
- `_getIconUrl` concatenates `imagePath` and icon URL option.

New situation: (i.e. with current PR)
- Individual icon path + file name retrieved from CSS. Each could be in arbitrary folder / URL. Recorded in icon URL option, unless it is already provided.
- `options.imagePath` used to refactor the above retrieved URL (currently just in case icon URL option is not provided), replacing the entire path except for the file name.
- `_getIconUrl` just outputs icon URL option as recorded (and possibly refactored).

Therefore the behaviour changes in case both `imagePath` and non-default icon URL options are provided:
- Currently: `_getIconUrl` concatenates `imagePath` and icon URL option.
- With this PR in current state: `_getIconUrl` just outputs icon URL option.

While the fix may look simple, actually it becomes complicated when we try to keep the possibility to specify a path in icon URL option as well (therefore we have to concatenate `imagePath`, we cannot replace the path), and we take into account the fact that each icon URL in CSS may have a completely different URL (therefore there may be no _common_ folder to record into `imagePath`).

Maybe some intermediate solution could be to still retrieve full URL from CSS if no `imagePath` is provided, and in the case it _is_ provided, strip the path before the file name so that it can be replaced by `imagePath`, and `_getIconUrl` works as currently in that case?

ghybs commented:
Ok should be fixed now:

- If `imagePath` option is specified, `IconDefault` emulates how it currently works: only the file name is retrieved from CSS and recorded in `iconUrl` option (same for shadow and Retina). Then `_getIconUrl` appends `imagePath` and `iconUrl`, so that it also works when the latter is provided by the application (and it can even contain a partial path).

- If no `imagePath` option is provided, the full icon URL is retrieved from CSS and recorded in options, so that they can have any arbitrary URL.

Added 2 new dedicated tests.

ghybs commented on src/layer/marker/Icon.Default.js:
```diff
@@ -19,42 +19,133 @@ import * as DomUtil from '../../dom/DomUtil';
+
+// Retrieve icon option values from CSS (icon or shadow).
+function _detectIconOptions(className, imagePath) {
+	var el = DomUtil.create('div',  className, document.body),
```
Remove 3rd argument `document.body`, otherwise the dummy element will actually be added to the document…

ghybs commented on src/layer/marker/Icon.Default.js:
```diff
@@ -19,42 +19,133 @@ import * as DomUtil from '../../dom/DomUtil';
+
+// Retrieve anchor option values from CSS (popup or tooltip).
+function _detectDivOverlayOptions(className) {
+	var el = DomUtil.create('div', className, document.body),
```
Remove 3rd argument `document.body`, otherwise the dummy element will actually be added to the document…
